#167165055 Send verification email

VictorMaria opened the pull request:
### What does this PR do?
Creates a functionality that sends verification link via email to a user upon successful registration

#### Description of Task to be completed?
- Given that a user successfully signs up
- The user is sent a verification link via email
- Upon clicking the link, the user is verified 

#### How should this be manually tested?
1. Fetch this branch on the terminal using 
`git fetch origin ft-send-verification-email-167165055`
2. Install dependencies using `npm install`
3. Start the server with `npm run server`
4. On Postman, sign up by making a POST request to /api/v1/auth/signup.
5. Check the inbox of the email you signed up with and click the link sent.
6. Upon redirection to the browser, copy the link in the URL bar.
7. Paste this link in Postman and make a PATCH request.


#### What are the relevant pivotal tracker stories?
[#167165055](https://www.pivotaltracker.com/story/show/167165055)

#### Screenshots (if appropriate)
<img width="820" alt="Screen Shot 2019-07-27 at 9 36 07 AM" src="https://user-images.githubusercontent.com/41482184/61992299-3dacff80-b054-11e9-84dc-b259a5fed1ee.png">
<img width="1157" alt="Screen Shot 2019-07-27 at 9 50 20 AM" src="https://user-images.githubusercontent.com/41482184/61992313-659c6300-b054-11e9-8622-872ecd774844.png">

Commit:
chore(signup): add user signup controller

Installed Jsonwebtoken for authentication
Installed bcryptjs for hashing password and user verification
stopped the app from using session
[Started #167165054]

Commit:
feat(signup):add tests to signup endpoint

Installed @babel/polyfill to resolve regenerator runtime error
Added tests to the signup endpoints

[Starts #167165054]

Commit:
feat(signup): add database on travis

added database on travis to make my test pass

[Starts #167165054]

VictorMaria committed:
feature(verification): send verification email
[Starts #167165055]

VictorMaria committed:
feature(verification): send verification email
[Starts #167165055]

VictorMaria committed:
feature(verification):send verification email
create functionality to send verification email upon sign up
create endpoint to verify user after verification link is clicked
add tests for the endpoint created
add Postals and EmailVerification js files
[Finishes #167165055]

VictorMaria committed:
feature(verification):send verification email
create functionality to send verification email upon sign up
create endpoint to verify user after verification link is clicked
add tests for the endpoint created
add Postals and EmailVerification js files
[Finishes #167165055]

mekzy-o committed:
feat(signup): Implement TTL corrections

Added Documentation on swagger.json

[Finishes #167165054]

VictorMaria committed:
rebase branch
[Finishes #167165055]

VictorMaria committed:
Merge branch 'ft-send-verification-email-167165055' of https://github.com/andela/mazus-ah-backend into ft-send-verification-email-167165055

rebase with branch ft-user-signup-167165054
modify EmailVerification and AuthController js files
make verification link dynamic for both local development and production
implement hound feedback
[Finishes #167165055]

VictorMaria committed:
resolve merge conflicts

VictorMaria committed:
Rebase with develop
Resolve merge conflicts
[Finishes 167165055]

tolulope-od commented on controllers/AuthController.js:
```diff
@@ -43,14 +41,21 @@ export default class AuthController {
+    const registeredUser = await models.User.create(user);
     const token = Helper.createToken({
       id: registeredUser.id,
-      email,
```
Hey, could you please add the trailing commas back here? thanks

VictorMaria replied:
I will see to that.

tolulope-od commented on helpers/EmailVerification.js:
```diff
@@ -0,0 +1,36 @@
+  * @memberof EmailVerification
+  * @returns {function} - returns a function call
+  */
+  static async sendVerificationEmail({ protocol, hostname }, recipient, name, token) {
```
Beautiful destructuring here! I love it!

VictorMaria replied:
Thank you very much

tolulope-od reviewed (commented):
Great work Victor I learned a lot just going through this PR. Could you please check the comments I left? Thanks

VictorMaria committed:
feature(verification): send verification email
Document verify email endpoint
write two tests to improve test coverage
put back trailing comma in authController js file
put objects containing mock users to be verified in mockUsersToVerify js file
[Finishes #167165055]

VictorMaria committed:
rebase branch with develop
resolve merge conflicts
[Finishes #167165055]

VictorMaria committed:
put back trailing commas in authController js file
[Finishes #167165055]

VictorMaria committed:
modify app.json content

VictorMaria committed:
fix test coverage drop
[Finishes #167165055]

VictorMaria committed:
Fix test coverage

VictorMaria committed:
improve text coverage

VictorMaria committed:
improve text coverage

VictorMaria committed:
Merge branch 'develop' of https://github.com/andela/mazus-ah-backend into ft-send-verification-email-167165055

VictorMaria committed:
Merge branch 'ft-send-verification-email-167165055' of https://github.com/andela/mazus-ah-backend into ft-send-verification-email-167165055

fix test

mazma1 commented on __tests__/authRoutes.test.js:
```diff
@@ -45,6 +47,69 @@ describe('User signup tests', () => {
+        done();
+      });
+  });
+});
```
Review the statements in your `describe` and `it` blocks. They should always form a coherent sentence that explains what each test case does.

mazma1 commented on __tests__/authRoutes.test.js:
```diff
@@ -4,13 +4,15 @@ import chaiHttp from 'chai-http';
 import app from '..';
 import models from '../database/models';
 import mockUsers from './mockData/mockUsers';
+import mockUsersToVerify from './mockData/mockUsersToVerify';
```
Destructure the values coming from `mockData`

VictorMaria replied:
I can't use the values coming from mockUsers because their emails would have been signed up by the time I need to run the signup before hooks for my test. That is why I created my own mockUsersToVerify

mazma1 commented on __tests__/mockData/mockUsersToVerify.js:
```diff
@@ -0,0 +1,21 @@
+    password: 'P455w0rd'
+  }
+};
+export default mockUsersToVerify;
```
Be consistent with trailing commas in this file

mazma1 commented on middlewares/inputValidation.js:
```diff
@@ -1,4 +1,4 @@
-import { check, validationResult } from 'express-validator';
+import { check, validationResult } from 'express-validator/check';
```
Do you have the latest changes from `develop`? Your copy of this file seem to be outdated

VictorMaria replied:
I had the latest develop as at that time

VictorMaria committed:
implement ttl feedback

VictorMaria committed:
Merge branch 'develop' into ft-send-verification-email-167165055

VictorMaria committed:
Merge branch 'develop' into ft-send-verification-email-167165055
